fix(core): add binding name to the error message `expression changed after it was checked`

brachi-wernick opened the pull request:
error `ExpressionChangedAfterItHasBeenCheckedError` doesn't provide which  binding is problematic, added this information to the error message

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
the error message just say the current value vs the previous value, for example:
`ExpressionChangedAfterItHasBeenCheckedError: Expression has changed after it was checked. Previous value: 'abc'. Current value: 'xyz'.`

Issue Number: N/A


## What is the new behavior?
the error message will say also what is the field name, for example:
`ExpressionChangedAfterItHasBeenCheckedError: Expression 'name' has changed after it was checked. Previous value: 'abc'. Current value: 'xyz'.`

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information
same as https://github.com/angular/angular/pull/17525 that wasn't touched since Jul12 and has too many changed files: 911 :open_mouth:

Commit:
fix(core): add binding name to the error message `expression changed after it was checked`

error `ExpressionChangedAfterItHasBeenCheckedError` doesn't provide which  binding is problematic, added this information to the error message

rossgardt commented on packages/core/src/view/errors.ts:
```diff
@@ -10,9 +10,9 @@ import {ERROR_DEBUG_CONTEXT, ERROR_LOGGER, getDebugContext} from '../errors';
 
 export function expressionChangedAfterItHasBeenCheckedError(
-    context: DebugContext, oldValue: any, currValue: any, isFirstCheck: boolean): Error {
+    context: DebugContext, oldValue: any, currValue: any, isFirstCheck: boolean,name:string): Error {
```
Small remark: spacings are missing `[..] boolean, name: string`.

Commit:
fix(core): format and null safe

format the code and make the function to be able get null binding name

Commit:
fix(core): fix test

error `ExpressionChangedAfterItHasBeenCheckedError` has now the binding field name, fix testing accordingly

Commit:
fix(core): fix test

error `ExpressionChangedAfterItHasBeenCheckedError` has now the binding field name, fix testing accordingly

brachi-wernick commented:
I signed it!

Commit:
fix(core): revert import

revert import

brachi-wernick commented:
@googlebot I signed it!

brachi-wernick commented:
@alexeagle or someone else with approval access:
I'll appreciate your help here: 
1. I have no idea why integration job fail, seems like a connection issue. can you suggest me what to do?
2. I'm still labeled as `cla:no` while I already singed it, and comment to the bot.

rossgardt commented on packages/core/src/view/errors.ts:
```diff
@@ -10,9 +10,11 @@ import {ERROR_DEBUG_CONTEXT, ERROR_LOGGER, getDebugContext} from '../errors';
+  let expression = name ? ` '${name}'` : '';
   let msg =
-      `ExpressionChangedAfterItHasBeenCheckedError: Expression has changed after it was checked. Previous value: '${oldValue}'. Current value: '${currValue}'.`;
+      `ExpressionChangedAfterItHasBeenCheckedError: Expression${expression} has changed after it was checked. Previous value: '${oldValue}'. Current value: '${currValue}'.`;
```
Please add a space in between `[..] Expression${expression} has changed`
(Sorry for being the annoying _spacing police_ 😅)

brachi-wernick replied:
@rossgardt it's OK to comment on spacing. But here I add the space to the `expression` value:
```typescript
let expression = name ? ` '${name}'` : '';
```
see that I want the space only if I have defined name. Other cases I want it to become empty string (no space). And then I will have a message without 2 spaces...
In case no name I want this msg to appear:
`Expression has changed...`
And not this:(double spaces)
`Expression  has changed...`

rossgardt commented:
Okay, I see. Nothing said 😉
On Tue 7. Nov 2017 at 07:09, brachi-wernick <notifications@github.com>
wrote:

> *@brachi-wernick* commented on this pull request.
> ------------------------------
>
> In packages/core/src/view/errors.ts
> <https://github.com/angular/angular/pull/20054#discussion_r149278226>:
>
> >    let msg =
> -      `ExpressionChangedAfterItHasBeenCheckedError: Expression has changed after it was checked. Previous value: '${oldValue}'. Current value: '${currValue}'.`;
> +      `ExpressionChangedAfterItHasBeenCheckedError: Expression${expression} has changed after it was checked. Previous value: '${oldValue}'. Current value: '${currValue}'.`;
>
> @rossgardt <https://github.com/rossgardt> it's OK to comment on spacing.
> But here I add the space to the expression value:
>
> let expression = name ? ` '${name}'` : '';
>
> see that I want the space only if I have defined name. Other cases I want
> it to become empty string (no space). And then I will have a message
> without 2 spaces...
> In case no name I want this msg to appear:
> Expression has changed...
> And not this:(double spaces)
> Expression has changed...
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/20054#discussion_r149278226>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAOVWP1Bgrup0XWikP3rx1Xrs9Zp2uRTks5sz_KJgaJpZM4QMuT2>
> .
>

mhevery commented:
@brachi-wernick it looks like @googlebot still thinks you have not signed it. Does your email you signed it with match the email on the PR? Can you resign with emails that match. Thanks

brachi-wernick commented:
@googlebot I signed it!

brachi-wernick commented:
@mhevery I fixed the issue with the `cla:no`,  I typed another mail, when signed in.
can you/someone else help me out with the integration job? it failed but seems like it isn't related to my changes.

brachi-wernick commented:
already done and merge in another PR https://github.com/angular/angular/pull/20352
I close this.
